Testing-the-engine.md

```diff
@@ -0,0 +1,302 @@
+are mocked by Robolectric. If you need to modify how Android components (such as
+an [android.view.View](https://developer.android.com/reference/android/view/View.html)
+or an [android.app.Activity](https://developer.android.com/reference/android/app/Activity.html))
+behave in the test, see other tests for examples or see docs at
+http://robolectric.org/ on how to interact with shadows.
+
+#### Mockito
+
+[Mockito](https://site.mockito.org/) is also a standard Android testing library
+used to mock non-Android dependencies needed to construct and test interactions
+with your your under-test production class.
+
+It's best practice to test only one real production class per test and
+mock all other dependencies with mockito.
+
+The Mockito library is an available test dependency when writing Robolectric
+tests.
+
+### Component integration tests
+
+Component tests test the interaction of multiple embedding Java classes together
+but doesn't test all production classes end-to-end. In the Android embedding
+case, we test groups of Java classes by their function in `...ComponentTest.java`
+files that are also in the `shell/platform/android/test/io/flutter/`
+directory. C++ engine parts via JNI are not tested here.
+
+Component tests are also Robolectric JUnit tests and are invoked together with
+unit tests when running:
+
+    `testing/run_tests.py --type=java`
+
+JUnit component tests are executed during pre-submit on our CI system when
+submitting PRs to the `flutter/engine` repository.
+
+### End-to-end tests
+
+End-to-end tests exercise the entire Android embedding with the C++ engine on
+a real Android runtime in an emulator. It's an integration test ensuring that
+the engine as a whole on Android is functioning correctly.
+
+The project containing the Android end-to-end engine test is at
+https://github.com/flutter/engine/tree/master/testing/scenario_app/android.
+
+This test project is build similarly to a normal Flutter app. With the Dart code
+compiled into AOT and the Android part compiled via Gradle with a dependency
+on the prebuilt local engine. It's then installed and executed on an emulator.
+
+Unlike a normal Flutter app, the Flutter framework on the Dart side is a
+lightweight fake at https://github.com/flutter/engine/tree/master/testing/scenario_app/lib
+that implements some of the basic functionalities of `dart:ui` Window rather
+than using the real Flutter framework at `flutter/flutter`.
+
+The end-to-end test can be executed by running:
+
+    `testing/scenario_app/build_and_run_android_tests.sh`
+
+Additional end-to-end instrumented tests can be added to https://github.com/flutter/engine/tree/master/testing/scenario_app/android/app/src/androidTest/java/dev/flutter/scenarios.
+
+If supporting logic are needed for the test case, they can be added to the
+Android app under-test in https://github.com/flutter/engine/tree/master/testing/scenario_app/android/app/src/main/java/dev/flutter/scenarios
```

Comment:
Terminology: "under-test"

Reply:
What did you mean?

Reply:
The term "under-test" here is a stretch since the testing setup is a bit weird.  I think it confuses the point more than helps.